messaging/src/main/java/org/axonframework/commandhandling/gateway/DefaultReactorCommandGateway.java

```diff
@@ -0,0 +1,253 @@
+     * it is.
+     * </p>
+     *
+     * @param builder the {@link Builder} used to instantiate a {@link DefaultReactorCommandGateway} instance
+     */
+    protected DefaultReactorCommandGateway(Builder builder) {
+        builder.validate();
+        this.commandBus = builder.commandBus;
+        this.retryScheduler = builder.retryScheduler;
+        this.dispatchInterceptors = builder.dispatchInterceptors;
+        this.resultInterceptors = builder.resultInterceptors;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link DefaultReactorCommandGateway}.
+     * <p>
+     * The {@code dispatchInterceptors} are defaulted to an empty list.
+     * The {@code resultHandlerInterceptors} are defaulted to an empty list.
+     * The {@link CommandBus} is a <b>hard requirement</b> and as such should be provided.
+     * </p>
+     *
+     * @return a Builder to be able to create a {@link DefaultReactorCommandGateway}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public <R> Mono<R> send(Object command) {
+        //noinspection unchecked
+        return Mono.<CommandMessage<?>>just(GenericCommandMessage.asCommandMessage(command))
+                .transform(this::processCommandInterceptors)
+                .flatMap(this::dispatchCommand)
+                .flatMap(this::processResultsInterceptors)
+                .transform(this::getPayload);
+    }
+
+    @Override
+    public Registration registerDispatchInterceptor(ReactorMessageDispatchInterceptor<CommandMessage<?>> interceptor) {
+        dispatchInterceptors.add(interceptor);
+        return () -> dispatchInterceptors.remove(interceptor);
+    }
+
+    @Override
+    public Registration registerResultHandlerInterceptor(
+            ReactorResultHandlerInterceptor<CommandMessage<?>, CommandResultMessage<?>> interceptor) {
+        resultInterceptors.add(interceptor);
+        return () -> resultInterceptors.remove(interceptor);
+    }
+
+    private Mono<CommandMessage<?>> processCommandInterceptors(Mono<CommandMessage<?>> commandMessage) {
+        return Flux.fromIterable(dispatchInterceptors)
+                   .reduce(commandMessage, (command, interceptor) -> interceptor.intercept(command))
+                   .flatMap(Mono::from);
+    }
+
+    private <C, R> Mono<Tuple2<CommandMessage<C>, Flux<CommandResultMessage<? extends R>>>> dispatchCommand(
+            CommandMessage<C> commandMessage) {
+        ReactiveCallback<C, R> reactiveCallback = new ReactiveCallback<>();
+        CommandCallback<C, R> callback = reactiveCallback;
+        if (retryScheduler != null) {
+            callback = new RetryingCallback<>(callback, retryScheduler, commandBus);
+        }
+        commandBus.dispatch(commandMessage, callback);
+        return Mono.just(commandMessage).zipWith(Mono.just(Flux.from(reactiveCallback)));
```

Comment:
Unfortunately, it's not because of making Mono from Publisher, the common result interceptor signature makes this code complex.

If ReactiveCallback would extend Mono instead of Published we would still need to convert Mono to Flux because of result interceptor signature (intercepts flux of results).

Anyway will try to think if there is a way to improve interceptor signature to make this code more readable and less complex

Reply:
Thanks for the explanation here @stefanvozd. Think for now, this is ok; it's good to have a think about this if we can improve this.